File: src/test/java/es/leanmind/matchers/fixtures/ListMatchers.java
package es.leanmind.matchers.fixtures;

import org.assertj.core.api.Assertions;

import java.util.List;

public class ListMatchers {
    private final List<Integer> actual;

    public ListMatchers(List<Integer> actual) {
        this.actual = actual;
    }

    public void isExactly(Number... expected) {
        Assertions.assertThat(expected)
                .hasSize(actual.size())
                .hasSameElementsAs(actual);
    }

    public static ListMatchers assertThatList(List<Integer> list) {
        return new ListMatchers(list);
    }
}

File: tdd-kotlin/kotlin/src/test/kotlin/leanmind/examples/ListMatchers.kt
package leanmind.examples

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.lang.AssertionError

fun assertThatList(list: List<Number>): ListMatchers {
    return ListMatchers(list)
}

class ListMatchers(private val actual: List<Number>) {

    fun isExactly(vararg expected: Number) {
        assertThat(expected).hasSize(actual.size).hasSameElementsAs(actual)
    }
}

class ListMatchersShould {

    @Test
    fun compare_same_elements_in_both_lists() {
        assertThatList(listOf(1, 2)).isExactly(1, 2)

        assertThrows<AssertionError> {
            assertThatList(listOf(1, 2)).isExactly(1, 3)
        }
    }

}

